### PR TITLE
[Perf] Skip per-call mat_a/scales_a padding in cutlass FP8 blockwise GEMM

### PR DESCRIPTION
**Pull request by yuan-luo**
## Summary

The cutlass FP8 blockwise linear path (`cutlass_w8a8_block_fp8_linear_with_fallback`) quantizes the activation, then calls `fp8_blockwise_scaled_mm`, whose sgl-kernel wrapper pads `mat_a` and `scales_a` up to a multiple of 4 rows on every invocation. Each pad is a `torch.zeros` fill plus a `torch.cat`, so every GEMM emits 2 extra fill and 2 extra cat kernels. This PR allocates the quantization outputs with row count already aligned to 4 up front, so the wrapper's `pad_tensor()` short-circuits (`pad_rows == 0`) and those four memory kernels are never launched. The GEMM then runs on the padded buffer and the caller slices the output back to the real row count.

This is a memory-traffic and kernel-count reduction only. The GEMM math, the quantization kernel, and the numerical result are unchanged.

## Motivation

In a decode-heavy trace (Qwen3.5-122B-A10B-FP8, TP=4, NEXTN speculative decoding) the per-GEMM kernel sequence is `group_quant -> fill -> cat -> fill -> cat -> copy -> copy -> cutlass_fp8_gemm`. The two fill and two cat kernels come entirely from `pad_tensor()` padding `mat_a` (the quantized activation) and `scales_a` (its column-major scales) to a 4-row boundary. The cutlass SM90 blockwise kernel requires the A operand row count to be a multiple of 4, and during speculative decode M is just 1 to 5 (the draft token count), so this padding fires on essentially every call. Because the quant output buffers are freshly allocated on every call, the wrapper has to re-pad each time, even though the required alignment is known before quantization. Allocating aligned buffers in the quant step removes that work without touching the kernel.

## How it works

The alignment requirement cannot be removed, but it can be paid once, for free, at buffer-allocation time in the quant step, instead of as a separate fill plus cat pass before every GEMM. The diagram below contrasts the two call flows: in the legacy path the wrapper rebuilds padded copies of `mat_a` and `scales_a` on each call; in this PR the quant step allocates 4-row-aligned buffers, so the wrapper's `pad_tensor()` sees `pad_rows == 0` and skips the fill and cat entirely.

image

Correctness follows from the memory layout in the bottom row of the diagram. A blockwise FP8 GEMM is row-independent along M, so the result of row `i` depends only on input row `i`. The legacy path appends freshly zeroed rows to reach `m_pad`; this PR allocates `m_pad` rows and leaves the tail uninitialized (garbage). Either way the first `m` rows are computed from identical inputs, and the output is sliced back to `m`, so the garbage rows are never read and cannot affect the result.

## Modifications

`python/sglang/srt/layers/quantization/fp8_kernel.py`: adds `sglang_per_token_group_quant_fp8_row_padded`, which allocates `x_q` as `(m_pad, k)` and the column-major `x_s` as `(m_pad, k // group)` where `m_pad = ceil_align(m, 4)`, runs the existing v2 quant kernel on the `[:m]` slices, and returns the full padded buffers. The `x_s` buffer is built as `(k // group, m_pad)` then transposed so that `stride(0) == 1`, which is the M-major layout the kernel requires for `scales_a`. Rows in `[m, m_pad)` are left uninitialized on purpose. When the v2 kernel is unavailable (group size not in {16, 32, 64, 128}) it falls back to the existing `sglang_per_token_group_quant_fp8` and lets the wrapper pad as before.

`python/sglang/srt/layers/quantization/fp8_utils.py`: `cutlass_w8a8_block_fp8_linear_with_fallback` now calls the row-padded quant and slices the GEMM output back to the original row count (`output[: input_2d.shape[0]]`) before adding bias. `weight_scale.T` is still passed as-is: the kernel requires `scales_b.stride(0) == 1` (a K-major view) and materializes its own contiguous copy internally, so that one is out of scope here.

## Correctness

A blockwise FP8 GEMM is row-independent along M, so computing extra padding rows cannot change the first `m` output rows; the uninitialized tail is sliced off and never read. A direct unit comparison against the existing unpadded path is bit-exact (`max_abs_diff == 0`) for `M` in {1, 2, 3, 4, 5, 7, 13, 64} on H20-3e (SM90), covering both already-aligned and unaligned M. End to end, a greedy 512-token completion is byte-for-byte identical before and after the change.

## Performance

Measured on H20-3e, TP=4, Qwen3.5-122B-A10B-FP8, NEXTN speculative decoding (num-steps 4, eagle-topk 1, draft-tokens 5), CUDA graph on, radix cache off.

The figure below is from two torch profiler traces captured under the same workload (310 decode steps each). Panel A shows per-step GPU time by kernel family, panel B shows memory-op launches per FP8 GEMM.
image

Decode-loop kernel counts from the traces:

| Kernel family | Baseline | This PR |
|---|---|---|
| cutlass FP8 GEMM | 23.4k launches, 454 ms | 23.4k launches, 454 ms |
| group quant | 35.0k launches, 107 ms | 35.0k launches, 107 ms |
| fill | ~65k launches, ~69 ms | 315 launches, 0.4 ms |
| cat | ~63k launches, ~89 ms | 364 launches, 0.7 ms |

End-to-end:

| Workload | Baseline | This PR | Delta |
|---|---|---|---|
| bs=1 single request, latency | 2.21 s | 2.21 s | no change |
| concurrency 64, output throughput | 2617 tok/s | 2671 tok/s | +2.1% |

The fill and cat kernels are essentially eliminated from the decode loop (about 158 ms of GPU time on this trace). The end-to-end gain is small and only shows up in the GPU-bound regime: a single bs=1 request is CPU-bound on speculative-decode scheduling, so removing GPU-side memory kernels off the critical path leaves latency unchanged. Under concurrency 64 the GPU is more saturated and the saved kernels translate into a measured +2.1% throughput, stable across four runs with no overlap against the baseline runs.

## Test plan

- [x] Unit: `test/registered/quant/test_fp8_blockwise_row_padding.py` asserts the row-padded buffers are 4-row aligned with M-major scales, the live rows match the legacy quant element-for-element, the full linear is `torch.equal` to the legacy unpadded GEMM for M in {1, 2, 3, 4, 5, 7, 13, 16, 31, 64, 256}, and a bf16 reference sanity bound. 3/3 pass on H20-3e (SM90); a negative control that drops the output slice fails as expected.
- [x] E2E: greedy 512-token completion byte-identical before and after.
- [x] Throughput: concurrency 64, 256 requests, +2.1% output tok/s, four runs each, non-overlapping.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #27332192942](https://github.com/sgl-project/sglang/actions/runs/27332192942)
Latest PR Test (Extra): :white_check_mark: [Run #27332221473](https://github.com/sgl-project/sglang/actions/runs/27332221473)

### CODE REVIEW

File: python/sglang/srt/layers/quantization/fp8_kernel.py [hunk 1]
```diff
@@ -565,6 +565,61 @@ def sglang_per_token_group_quant_fp8(
...
+    multiple of 4 rows on every call (a zeros fill + a cat for each of mat_a
+    and scales_a). Allocating the quant outputs with rows already aligned to
+    ``row_alignment`` makes the wrapper's pad_tensor() short-circuit (pad_rows
+    == 0), removing 2x fill + 2x cat kernels per GEMM. Rows in [m, m_pad) are
+    uninitialized garbage; the caller must slice the GEMM output back to m.
+    """
+    assert x.dim() == 2, "row-padded quant expects a 2D input"
+    assert (
+        x.shape[-1] % group_size == 0
+    ), "the last dimension of `x` cannot be divisible by `group_size`"
```
[the remark this answers is not in the document]
Comment (yuan-luo) [reply python/sglang/srt/layers/quantization/fp8_kernel.py]: Good catch, fixed.

**Review by b8zhong — APPROVED** [review]
